---
title: "BW_Files_Analysis_Project"
author: "Florence Marti"
date: "2025-06-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd("~/Documents/GitHub/EpiGen_Project_Valerie_Florence/Histone Modifications Memory")
```

```{r, include = FALSE}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(ensembldb)
  library(GenomicRanges)
  library(epiwraps)
  library(rtracklayer)
  library(ggplot2)
  library(magick)
  library(R.utils) 
  library(MASS)
  library(org.Mm.eg.db)
  library(AnnotationDbi)
  library(valr)
  library(EnrichedHeatmap)
  library(gridExtra)
  library(grid)
  library(motifmatchr)
  library(MotifDb)
  library(universalmotif)
  library(PWMEnrich)
})
```

```{r}
ah <- AnnotationHub(localHub=TRUE)
ensdb <- ah[["AH89211"]] # mouse ensembldb object
genes <- genes(ensdb)
```

We have hPMTs and one motif (CTCF), we could try to motif match??

```{r}

```


# Comparing the histone modifications of naive T cells
First reinporting the GRanges files of all the peaks
```{r}
peaks_part1 <- GRangesList(
  H3K27ac = import("peak_bed_files/H3K27ac_peaks.bed"),
  H3K27me3 = import("peak_bed_files/H3K27me3_peaks.bed"),
  H3K4me3 = import("peak_bed_files/H3K4me3_peaks.bed"),
  H3K9me3 = import("peak_bed_files/H3K9me3_peaks.bed")
)

peaks_open_me_na <- read.csv("files/memory_open_peaks_signal.csv", header = T)
peaks_closed_me_na <- read.csv("files/memory_closed_peaks_signal.csv", header = T)
peaks_biv_me_na <- read.csv("files/memory_bivalent.csv", header = T)
```



```{r}
bw_files <- list.files("bigwig_files", pattern = ".\\.bw$", full.names = T)

names(bw_files) <- gsub("\\.bw$", "", basename(bw_files))
```


```{r}
# Naive
bwfiles_naive <- list.files("bigwig_files", pattern = "_naive.bw$", full.names = TRUE)
names(bwfiles_naive) <- gsub("\\.bw$", "", basename(bwfiles_naive))

naive_regions <- reduce(unlist(GRangesList(peaks_part1)))
sm_n <- signal2Matrix(bwfiles_naive, regions = naive_regions, w = 40)


# Memory
bwfiles_memory <- list.files("bigwig_files", pattern = "_memory.bw$", full.names = TRUE)
names(bwfiles_memory) <- gsub("\\.bw$", "", basename(bwfiles_memory))

memory_regions <- reduce(unlist(GRangesList(peaks_part1)))
sm_m <- signal2Matrix(bwfiles_memory, regions = memory_regions, w = 40)

# Exhausted

bwfiles_exh <- list.files("bigwig_files", pattern = "_EX.bw$", full.names = TRUE)
names(bwfiles_exh) <- gsub("\\.bw$", "", basename(bwfiles_exh))

exh_regions <- reduce(unlist(GRangesList(peaks_part1)))
sm_e <- signal2Matrix(bwfiles_exh, regions = exh_regions, w = 40)

```

```{r}
plotEnrichedHeatmaps(sm_n, multiScale = T)
plotEnrichedHeatmaps(sm_m,  multiScale = T)
plotEnrichedHeatmaps(sm_e,  multiScale = T)
```

```{r}
bivalent_overlaps <- peaks_part1$H3K27me3[overlapsAny(peaks_part1$H3K27me3, peaks_part1$H3K4me3)]

sm2 <- signal2Matrix(bw_files, regions = bivalent_overlaps, w = 50)

colData(sm2)
```

Visualising a single region (w6)
```{r}
plotSignalTracks(files = bw_files, region = peaks_part1$H3K27ac[1], extend=5000, tracks.params = list(ylim=c(0,170)), 
                  ensdb = ensdb,  transcripts = "full" )
plotSignalTracks(files = bw_files, region = peaks_part1$H3K4me3[1], extend=5000, tracks.params = list(ylim=c(0,170)), 
                  ensdb = ensdb,  transcripts = "full" )
plotSignalTracks(files = bw_files, region = peaks_part1$H3K27me3[1], extend=5000, tracks.params = list(ylim=c(0,170)), 
                  ensdb = ensdb,  transcripts = "full" )
plotSignalTracks(files = bw_files, region = peaks_part1$H3K9me3[1], extend=5000, tracks.params = list(ylim=c(0,170)), 
                  ensdb = ensdb,  transcripts = "full" )
```

Generating coverage tracks
```{r}

```




```{r}
```
Comparing the enriched heatmaps of all of the peaks and the peaks we filtered out as significant

```{r}
plotEnrichedHeatmaps(sm_n, trim = 0.05)
```


```{r}
plotEnrichedHeatmaps(sm_n_filtered, trim = 0.05)
```
```{r}
plotEnrichedHeatmaps(sm_m, trim=0.05)
```


```{r}
plotEnrichedHeatmaps(sm_m, trim=0.05)

meanEnrichment <- rowMeans(score(sm_m))
sm_m_2 <- sm_m[which(meanEnrichment >= median(meanEnrichment)),]

plotEnrichedHeatmaps(sm_m_2)
```

