---
title: "Project_2_Filtering_and_Annotation"
author: "Florence Marti"
date: "2025-06-11"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
setwd("~/Documents/GitHub/EpiGen_Project_Valerie_Florence")
```

```{r, include = FALSE}
suppressPackageStartupMessages({
  library(AnnotationHub)
  library(ensembldb)
  library(GenomicRanges)
  library(epiwraps)
  library(rtracklayer)
  library(ggplot2)
  library(magick)
  library(R.utils) 
  library(MASS)
  library(org.Mm.eg.db)
  library(AnnotationDbi)
  library(valr)
  library(EnrichedHeatmap)
  library(gridExtra)
  library(grid)
  library(limma)
  library(plyranges)
  library(dplyr)
})
```

# 6.)  Data Processing
```{r, include = T}
# Setting my heatmap colour right in the beginning
my_breaks <- seq(-2, 2, length.out = 101)
```

## 6.1) Importing the downloaded data

```{r, tab.cap="Table 1. Overview of contents of txt files available for download."}
head(read.table("peaks/H3K27ac.txt", header = TRUE))
```
In our txt files we find the information about the peak number, the chromosome where the peak is located, the start and end base, as well as the length.The file includes the normalised counts for all four replicates of our four (Naïve D30, Naïve D32, Memory, Exhausted) T-cell types. 

### 6.1.1) Conversion of .txt files to .bed files

```{r}

#CTCF_df <- read.table("peaks/CTCF.txt", header = TRUE)
H3K27ac_df <- read.table("peaks/H3K27ac.txt", header = TRUE)
H3K27me3_df <- read.table("peaks/H3K27me3.txt", header = TRUE)
H3K4me3_df <- read.table("peaks/H3K4me3.txt", header = TRUE)
H3K9me3_df <- read.table("peaks/H3K9me3.txt", header = TRUE)

# Columns for D32 Naïve T-cells were excluded for comparability of the results. 

H3K27ac_df <- H3K27ac_df[, -c(11:14)]
H3K27me3_df <- H3K27me3_df[, -c(11:14)]
H3K4me3_df <- H3K4me3_df[, -c(11:14)]
H3K9me3_df <- H3K9me3_df[, -c(11:14)]

# Converting the txt files to GRanges files

H3K27ac_gr <- GRanges(seqnames = H3K27ac_df$Chr, ranges = IRanges(start = H3K27ac_df$Start, end = H3K27ac_df$End), peak = H3K27ac_df$peakN)
H3K27me3_gr <- GRanges(seqnames = H3K27me3_df$Chr, ranges = IRanges(start = H3K27me3_df$Start, end = H3K27me3_df$End), peak = H3K27me3_df$peakN)
H3K4me3_gr <- GRanges(seqnames = H3K4me3_df$Chr, ranges = IRanges(start = H3K4me3_df$Start, end = H3K4me3_df$End), peak = H3K4me3_df$peakN)
H3K9me3_gr <- GRanges(seqnames = H3K9me3_df$Chr, ranges = IRanges(start = H3K9me3_df$Start, end = H3K9me3_df$End), peak = H3K9me3_df$peakN)


# Extracting the peaks of all available 
peaks <- GRangesList(
  H3K27ac = H3K27ac_gr,
  H3K27me3 = H3K27me3_gr,
  H3K4me3 = H3K4me3_gr,
  H3K9me3 = H3K9me3_gr
)

dir.create("peak_bed_files")
rtracklayer::export(peaks$H3K27ac, "peak_bed_files/H3K27ac_peaks.bed")
rtracklayer::export(peaks$H3K27me3, "peak_bed_files/H3K27me3_peaks.bed")
rtracklayer::export(peaks$H3K4me3, "peak_bed_files/H3K4me3_peaks.bed")
rtracklayer::export(peaks$H3K9me3, "peak_bed_files/H3K9me3_peaks.bed")

```


```{r, fig.align = "center", fig.cap="Figure 1. UpSet plot showing the overlap of genomic regions (peaks) identified for each hPMT (H3K4me3, H3K27ac, H3K27me3, H3K9me3) in naïve, memory, and exhausted T-cells. The plot displays the intersection sizes of peak sets for individual and combined histone modifications. Horizontal bars on the left indicate the total number of peaks detected for each mark, while vertical bars represent the number of genomic regions shared among specific combinations of modifications, as indicated by the connected dots below each bar."}
regionUpset(peaks)
```

The UpSet plot reveals substantial overlap as well as unique sets of genomic regions marked by each histone modification. The largest intersection (13,671 regions) corresponds to peaks uniquely associated with H3K27ac, highlighting its broad distribution. Notably, thousands of regions are marked exclusively by individual modifications (e.g., 7,224 for H3K9me3 and 7,138 for H3K27me3), while smaller subsets of peaks are co-occupied by multiple marks, reflecting combinatorial chromatin states. These patterns underscore both the specificity and redundancy of histone modification landscapes in T-cells, with distinct and overlapping regulatory regions contributing to the epigenetic diversity observed across naïve, memory, and exhausted states.


## 6.2) Quality Control

### 6.2.1) PCA
A principal component analysis (PCA) was performed to see how the different cell types within the hPMT would cluster to visualise first differences between the samples and hPMTs. 

#### 6.2.1.1) Clustering by Cell Type

```{r}
H3K27ac_PCA <- H3K27ac_df[,7:18]
rownames(H3K27ac_PCA) <- H3K27ac_df$peakN

H3K4me3_PCA <- H3K4me3_df[,7:18]
rownames(H3K4me3_PCA) <- H3K4me3_df$peakN

H3K27me3_PCA <- H3K27me3_df[,7:18]
rownames(H3K27me3_PCA) <- H3K27me3_df$peakN

H3K9me3_PCA <- H3K9me3_df[,7:18]
rownames(H3K9me3_PCA) <- H3K9me3_df$peakN


```

```{r}
run_pca <- function(signal_matrix, mark_name) {
  pca_input <- t(signal_matrix)
  pca_result <- prcomp(pca_input, scale. = TRUE) # scale is important for z scoring
  
  variance <- summary(pca_result)$importance[2, ]
  pca_df <- data.frame(
    PC1 = pca_result$x[, 1],
    PC2 = pca_result$x[, 2],
    CellType = rep(c("Naïve", "Memory", "Exhausted"), each = 4)
  )
  
  ggplot(pca_df, aes(PC1, PC2, color = CellType)) +
  geom_point(size = 4,) +
  scale_color_manual(values = c("Naïve" = "#ef77b4", "Memory" = "#ff7f0e", "Exhausted" = "#4daf4a" )) +
    labs(
      title = paste("PCA of", mark_name, "Signal"),
      x = paste0("PC1 (", round(variance[1]*100, 1), "%)"),
      y = paste0("PC2 (", round(variance[2]*100, 1), "%)")
    ) +
  theme_classic()
}


plot_H3K27ac <- run_pca(H3K27ac_PCA, "H3K27ac")
plot_H3K4me3 <- run_pca(H3K4me3_PCA, "H3K4me3")
plot_H3K27me3 <- run_pca(H3K27me3_PCA, "H3K27me3")
plot_H3K9me3 <- run_pca(H3K9me3_PCA, "H3K9me3")
```

```{r, fig.align = "center", fig.cap = "Figure 2. PCA of normalised hPMT signals (H3K27ac, H3K4me3, H3K27me3, H3K9me3) across naïve, memory, and exhausted T-cell states. Each panel shows the distribution of samples by cell type for a specific histone mark, with points colored according to cell type. The axes represent the first two principal components, indicating the proportion of variance explained. Clear separation between cell types is observed for each modification, highlighting distinct epigenetic landscapes among naïve, memory, and exhausted T-cells"}
library(gridExtra)
grid.arrange(
  plot_H3K27ac, plot_H3K4me3,
  plot_H3K27me3, plot_H3K9me3,
  nrow = 2,
  top = "PCA of hPMTs Across all Cell Types"
)
```

The PCA plots in Figure 2 show clear segregation of naïve, memory, and exhausted T-cell samples based on their hPMT.  For each histone mark — H3K27ac, H3K4me3, H3K27me3, and H3K9me3 — samples cluster according to cell type. This consistent separation across all four hPMTs indicates that each T-cell state is characterised by a unique chromatin landscape. The tight clustering of replicates within each cell type further demonstrates the reproducibility of the data and the robust epigenetic differences that accompany T-cell differentiation and exhaustion.

#### 6.2.1.2) Clustering by hPMT

```{r}
H3K27ac_PCA <- H3K27ac_PCA %>% tibble::rownames_to_column("peakID")
H3K4me3_PCA <- H3K4me3_PCA %>% tibble::rownames_to_column("peakID")
H3K27me3_PCA <- H3K27me3_PCA %>% tibble::rownames_to_column("peakID")
H3K9me3_PCA <- H3K9me3_PCA %>% tibble::rownames_to_column("peakID")

combined_signals <- list(
  H3K27ac_PCA,
  H3K4me3_PCA,
  H3K27me3_PCA,
  H3K9me3_PCA
) %>% 
  Reduce(function(x, y) inner_join(x, y, by = "peakID"), .)

rownames(combined_signals) <- combined_signals$peakID
combined_signals$peakID <- NULL

pca_input <- t(combined_signals) 
pca_result <- prcomp(pca_input, scale. = TRUE)

annotation_col <- data.frame(
  CellType = rep(rep(c("Naïve", "Memory", "Exhausted"), each = 4), times = 4),
  HistoneMark = rep(c("H3K4me3", "H3K27ac", "H3K27me3", "H3K9me3"), each = 12)
)

rownames(annotation_col) <- colnames(combined_signals)

pca_df <- data.frame(
  PC1 = pca_result$x[, 1],
  PC2 = pca_result$x[, 2],
  CellType = annotation_col$CellType,
  HistoneMark = annotation_col$HistoneMark
)
```


```{r, fig.align = "center", fig.cap = "Figure 3. PCA of normalised signals for four histone marks (H3K27ac, H3K4me3, H3K27me3, H3K9me3) in naïve, memory, and exhausted T-cells. Each point represents a sample, colored by cell type and shaped by histone mark. The first two principal components (PC1 and PC2) explain 40.5% and 25.1% of the variance, respectively. The PCA shows clear separation of samples according to both hPMT and cell state, with distinct clustering patterns for each histone mark."}
# Plot with custom aesthetics

ggplot(pca_df, aes(PC1, PC2, color = CellType, shape = HistoneMark)) +
  geom_point(size = 4, alpha = 0.8) +
  scale_color_manual(values = c("Naïve" = "#ef77b4", "Memory" = "#ff7f0e", "Exhausted" = "#4daf4a")) +
  scale_shape_manual(values = c(16, 17, 15, 18)) +  # Different shapes for marks
  theme_classic() +
  labs(
    title = "PCA of hPMTs in T-cells",
    x = paste0("PC1 (", round(summary(pca_result)$importance[2,1]*100, 1), "%)"),
    y = paste0("PC2 (", round(summary(pca_result)$importance[2,2]*100, 1), "%)")
  ) +
  guides(color = guide_legend(order = 1), shape = guide_legend(order = 2))

```


This PCA shows that the global patterns of hPMTs distinctly separate both by mark and by T-cell state. Samples cluster tightly by hPMT (shape), and within each mark, further group by cell type (color). This indicates that both the choice of histone mark and the differentiation state of the T-cell are major drivers of epigenetic variation. The absence of overlap between hPMT driven clusters suggests robust and reproducible differences in chromatin state among naïve, memory, and exhausted T-cells. 

## 6.3) Differential hPMT Analysis in Naïve vs. Memory T-cells
Signal intensities were extracted from GRanges objects for each hPMT (H3K27ac, H3K4me3, H3K27me3, H3K9me3) across naïve, memory, and exhausted T-cell states. Group means were calculated per genomic peak, followed by log₂ fold change (log₂FC) computation. Peaks with |log₂FC| > 0.58496 (1.5-fold change) were retained. 

### 6.3.1) Creating the datasets for comparison

```{r}
# Naïve = cols 7:10, Memory = cols 11:14, Exhausted: 15:18
H3K27ac_signal <- cbind(
  naive = H3K27ac_df[, 7:10],
  memory = H3K27ac_df[, 11:14],
  exh = H3K27ac_df[, 15:18]
)
H3K27me3_signal <- cbind( 
  naive = H3K27me3_df[, 7:10],
  memory = H3K27me3_df[, 11:14],
  exh = H3K27me3_df[, 15:18]
)
H3K4me3_signal <- cbind(
  naive = H3K4me3_df[, 7:10],
  memory = H3K4me3_df[, 11:14],
  exh = H3K4me3_df[, 15:18]
)

H3K9me3_signal <- cbind(
  naive = H3K9me3_df[, 7:10],
  memory = H3K9me3_df[, 11:14],
  exh = H3K9me3_df[, 15:18]
)

rownames(H3K27ac_signal) <- paste(H3K27ac_df$peakN)
rownames(H3K27me3_signal) <- paste(H3K27me3_df$peakN)
rownames(H3K4me3_signal)  <- paste(H3K4me3_df$peakN)
rownames(H3K9me3_signal)  <- paste(H3K9me3_df$peakN)
```

### 6.3.2) Naïve and memory T-cells
In order to avoid log2(0), a small pseudocount was added.

```{r}
pseudocount <- 1e-6
## H3K4me3
mean_naive_H3K4me3 <- rowMeans(H3K4me3_signal[, 1:4])   # 40269 elements
mean_memory_H3K4me3 <- rowMeans(H3K4me3_signal[, 5:8])  
mean_exh_H3K4me3 <- rowMeans(H3K4me3_signal[, 9:12])

log2_H3K4me3 <- log2((mean_memory_H3K4me3 + pseudocount) / (mean_naive_H3K4me3 + pseudocount))
sig_rows_H3K4me3 <- which(abs(log2_H3K4me3) > 0.58496)
sig_H3K4me3_signal <- H3K4me3_signal[sig_rows_H3K4me3, 1:8] #14862 elements
scaled_signal_H3K4me3 <- t(scale(t(sig_H3K4me3_signal)))   

## H3K27ac
mean_naive_H3K27ac <- rowMeans(H3K27ac_signal[, 1:4])  
mean_memory_H3K27ac <- rowMeans(H3K27ac_signal[, 5:8])  
mean_exh_H3K27ac <- rowMeans(H3K27ac_signal[, 9:12])  

log2_H3K27ac <- log2((mean_memory_H3K27ac + pseudocount) / (mean_naive_H3K27ac + pseudocount))
sig_rows_H3K27ac <- which(abs(log2_H3K27ac) > 0.58496)
sig_H3K27ac_signal <- H3K27ac_signal[sig_rows_H3K27ac,1:8 ]
scaled_signal_H3K27ac <- t(scale(t(sig_H3K27ac_signal)))  

## H3K27me3
mean_naive_H3K27me3 <- rowMeans(H3K27me3_signal[, 1:4])   
mean_memory_H3K27me3 <- rowMeans(H3K27me3_signal[, 5:8]) 
mean_exh_H3K27me3 <- rowMeans(H3K27me3_signal[, 9:12])  

log2_H3K27me3 <- log2((mean_memory_H3K27me3 + pseudocount) / (mean_naive_H3K27me3 + pseudocount))
sig_rows_H3K27me3 <- which(abs(log2_H3K27me3) > 0.58496)
sig_H3K27me3_signal <- H3K27me3_signal[sig_rows_H3K27me3, 1:8 ]
scaled_signal_H3K27me3 <- t(scale(t(sig_H3K27me3_signal)))  


## H3K9me3
mean_naive_H3K9me3 <- rowMeans(H3K9me3_signal[, 1:4])   
mean_memory_H3K9me3 <- rowMeans(H3K9me3_signal[, 5:8]) 
mean_exh_H3K9me3 <- rowMeans(H3K9me3_signal[, 9:12])  

log2_H3K9me3 <- log2((mean_memory_H3K9me3 + pseudocount) / (mean_naive_H3K9me3 + pseudocount))
sig_rows_H3K9me3 <- which(abs(log2_H3K9me3) > 0.58496)
sig_H3K9me3_signal <- H3K9me3_signal[sig_rows_H3K9me3, 1:8 ]
scaled_signal_H3K9me3 <- t(scale(t(sig_H3K9me3_signal)))  
```

### 6.3.3) Naïve and exhausted T-cells:
```{r}
## H3K4me3
log2_H3K4me3_EXN <- log2((mean_exh_H3K4me3 + pseudocount) / (mean_naive_H3K4me3 + pseudocount))
sig_rows_H3K4me3_EXN <- which(abs(log2_H3K4me3_EXN) > 0.58496)
sig_H3K4me3_signal_EXN <- H3K4me3_signal[sig_rows_H3K4me3_EXN,c(1:4, 9:12)]
scaled_signal_H3K4me3_EXN <- t(scale(t(sig_H3K4me3_signal_EXN)))  


## H3K27ac
log2_H3K27ac_EXN <- log2((mean_exh_H3K27ac + pseudocount) / (mean_naive_H3K27ac + pseudocount))
sig_rows_H3K27ac_EXN <- which(abs(log2_H3K27ac_EXN) > 0.58496)
sig_H3K27ac_signal_EXN <- H3K27ac_signal[sig_rows_H3K27ac,c(1:4, 9:12)]
scaled_signal_H3K27ac_EXN <- t(scale(t(sig_H3K27ac_signal_EXN)))  

# H3K27me3
log2_H3K27me3_EXN <- log2((mean_exh_H3K27me3 + pseudocount) / (mean_naive_H3K27me3 + pseudocount))
sig_rows_H3K27me3_EXN <- which(abs(log2_H3K27me3_EXN) > 0.58496)
sig_H3K27me3_signal_EXN <- H3K27me3_signal[sig_rows_H3K27me3_EXN,c(1:4, 9:12)]
scaled_signal_H3K27me3_EXN <- t(scale(t(sig_H3K27me3_signal_EXN)))  

# H3K9me3
log2_H3K9me3_EXN <- log2((mean_exh_H3K9me3 + pseudocount) / (mean_naive_H3K9me3 + pseudocount))
sig_rows_H3K9me3_EXN <- which(abs(log2_H3K9me3_EXN) > 0.58496)
sig_H3K9me3_signal_EXN <- H3K9me3_signal[sig_rows_H3K9me3_EXN,c(1:4, 9:12)]
scaled_signal_H3K9me3_EXN <- t(scale(t(sig_H3K9me3_signal_EXN)))  
```


### 6.3.4)  Memory and exhausted T-cells:
```{r}
# H3K4me3
log2_H3K4me3_EXM<- log2((mean_exh_H3K4me3 + pseudocount) / (mean_memory_H3K4me3 + pseudocount))
sig_rows_H3K4me3_EXM<- which(abs(log2_H3K4me3_EXM) > 0.58496)
sig_H3K4me3_signal_EXM<- H3K4me3_signal[sig_rows_H3K4me3_EXM, 5:12 ]
scaled_signal_H3K4me3_EXM<- t(scale(t(sig_H3K4me3_signal_EXM)))  


# H3K27ac
log2_H3K27ac_EXM<- log2((mean_exh_H3K27ac + pseudocount) / (mean_memory_H3K27ac + pseudocount))
sig_rows_H3K27ac_EXM<- which(abs(log2_H3K27ac_EXM) > 0.58496)
sig_H3K27ac_signal_EXM<- H3K27ac_signal[sig_rows_H3K27ac, 5:12]
scaled_signal_H3K27ac_EXM<- t(scale(t(sig_H3K27ac_signal_EXM)))  

# H3K27me3
log2_H3K27me3_EXM<- log2((mean_exh_H3K27me3 + pseudocount) / (mean_memory_H3K27me3 + pseudocount))
sig_rows_H3K27me3_EXM<- which(abs(log2_H3K27me3_EXM) > 0.58496)
sig_H3K27me3_signal_EXM<- H3K27me3_signal[sig_rows_H3K27me3_EXM,5:12 ]
scaled_signal_H3K27me3_EXM<- t(scale(t(sig_H3K27me3_signal_EXM)))  

# H3K9me3
log2_H3K9me3_EXM<- log2((mean_exh_H3K9me3 + pseudocount) / (mean_memory_H3K9me3 + pseudocount))
sig_rows_H3K9me3_EXM<- which(abs(log2_H3K9me3_EXM) > 0.58496)
sig_H3K9me3_signal_EXM <- H3K9me3_signal[sig_rows_H3K9me3_EXM, 5:12]
scaled_signal_H3K9me3_EXM<- t(scale(t(sig_H3K9me3_signal_EXM)))  
```


## 6.4)  Visualisation of Cell Type Comparison

### 6.4.1) Naïve vs memory T-cell

```{r}
annotation_col <- data.frame(
  CellType = rep(c("Naïve T-cell", "Memory T-cell"), each = 4)
)

annotation_colors <- list(
  CellType = c("Naïve T-cell" = "#ef77b4", "Memory T-cell" = "#ff7f0e"))


rownames(annotation_col) <- colnames(scaled_signal_H3K4me3)  
hmap1 <- pheatmap::pheatmap(scaled_signal_H3K4me3, breaks = my_breaks,  
                            annotation_col = annotation_col,
                            annotation_colors = annotation_colors, 
                            show_colnames = FALSE, show_rownames = FALSE, silent = TRUE)

rownames(annotation_col) <- colnames(scaled_signal_H3K27ac)  
hmap2 <- pheatmap::pheatmap(scaled_signal_H3K27ac, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors,show_colnames = FALSE, show_rownames = FALSE, silent = TRUE)


rownames(annotation_col) <- colnames(scaled_signal_H3K27me3)  
hmap3 <- pheatmap::pheatmap(scaled_signal_H3K27me3, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors,show_colnames = FALSE, show_rownames = FALSE,  silent = TRUE)

rownames(annotation_col) <- colnames(scaled_signal_H3K9me3)  
hmap4 <- pheatmap::pheatmap(scaled_signal_H3K9me3, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors, show_colnames = FALSE, show_rownames = FALSE, silent = TRUE)
```


### 6.4.2) Naïve vs exhausted T-cell
```{r}
annotation_col <- data.frame(
  CellType = rep(c("Naïve T-cell", "Exhausted T-cell"), each = 4)
)

annotation_colors <- list(
  CellType = c("Naïve T-cell" = "#ef77b4", "Exhausted T-cell" = "#4daf4a" ))


rownames(annotation_col) <- colnames(scaled_signal_H3K4me3_EXN)  
hmap5 <- pheatmap::pheatmap(scaled_signal_H3K4me3_EXN, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors, show_colnames = FALSE, show_rownames = FALSE, silent = TRUE)


rownames(annotation_col) <- colnames(scaled_signal_H3K27ac_EXN)  
hmap6 <- pheatmap::pheatmap(scaled_signal_H3K27ac_EXN, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors,show_colnames = FALSE, show_rownames = FALSE, silent = TRUE)


rownames(annotation_col) <- colnames(scaled_signal_H3K27me3_EXN)  
hmap7 <- pheatmap::pheatmap(scaled_signal_H3K27me3_EXN, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors,show_colnames = FALSE, show_rownames = FALSE,  silent = TRUE)

rownames(annotation_col) <- colnames(scaled_signal_H3K9me3_EXN)  
hmap8 <- pheatmap::pheatmap(scaled_signal_H3K9me3_EXN, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors, show_colnames = FALSE, show_rownames = FALSE, silent = TRUE)
```


### 6.4.3) Memory vs exhausted T-cell
```{r}
annotation_col <- data.frame(
  CellType = rep(c("Memory T-cell", "Exhausted T-cell"), each = 4)
)

annotation_colors <- list(
  CellType = c("Memory T-cell" = "#ff7f0e", "Exhausted T-cell" = "#4daf4a" ))

rownames(annotation_col) <- colnames(scaled_signal_H3K4me3_EXM)  
hmap9 <- pheatmap::pheatmap(scaled_signal_H3K4me3_EXM, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors, show_colnames = FALSE, show_rownames = FALSE, silent = TRUE)


rownames(annotation_col) <- colnames(scaled_signal_H3K27ac_EXM)  
hmap10 <- pheatmap::pheatmap(scaled_signal_H3K27ac_EXM, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors,show_colnames = FALSE, show_rownames = FALSE, silent = TRUE)


rownames(annotation_col) <- colnames(scaled_signal_H3K27me3_EXM)  
hmap11 <- pheatmap::pheatmap(scaled_signal_H3K27me3_EXM, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors,show_colnames = FALSE, show_rownames = FALSE,  silent = TRUE)

rownames(annotation_col) <- colnames(scaled_signal_H3K9me3_EXM)  
hmap12 <- pheatmap::pheatmap(scaled_signal_H3K9me3_EXM, breaks = my_breaks,  annotation_col = annotation_col,annotation_colors = annotation_colors, show_colnames = FALSE, show_rownames = FALSE, silent = TRUE)
```

### 6.4.4) Heatmaps of z-scored signal intensities for significantly different peaks in each hPMT
```{r fig.align='center', fig.height=10, fig.width=12.5, fig.align = "center", fig.cap="Figure 4. Heatmaps of z-scored signal intensities for significantly different peaks in each hPMT. Each heatmap displays normalised (z-scored) signal intensities for genomic regions with significant differences (|log₂ fold change| > 0.58496) in histone modification between T-cell states: Naïve vs Memory (left), Naïve vs Exhausted (center), and Memory vs Exhausted (right). Rows represent individual genomic peaks, and columns represent biological replicates, annotated by cell type. Separate panels are shown for H3K27ac, H3K4me3, H3K27me3, and H3K9me3 (top to bottom). Distinct clustering patterns highlight widespread and coordinated changes in chromatin accessibility and repression associated with T-cell differentiation and exhaustion "}

label_grobs <- arrangeGrob(
  textGrob("H3K27ac", rot = 90, gp = gpar(fontsize = 12, fontface = "bold")),
  textGrob("H3K4me3", rot = 90, gp = gpar(fontsize = 12, fontface = "bold")),
  textGrob("H3K27me3", rot = 90, gp = gpar(fontsize = 12, fontface = "bold")),
  textGrob("H3K9me3", rot = 90, gp = gpar(fontsize = 12, fontface = "bold")),
  ncol = 1,
  heights = unit(rep(1, 4), "null")  # match heatmap row heights
)

col1 <- arrangeGrob(hmap2$gtable, hmap1$gtable, hmap3$gtable, hmap4$gtable, ncol = 1)
col2 <- arrangeGrob(hmap6$gtable, hmap5$gtable, hmap7$gtable, hmap8$gtable, ncol = 1)
col3 <- arrangeGrob(hmap10$gtable, hmap9$gtable, hmap11$gtable, hmap12$gtable, ncol = 1)

col1_with_title <- arrangeGrob(textGrob("Naïve vs Memory", gp = gpar(fontsize = 12, fontface = "bold")),
                               col1,
                               ncol = 1,
                               heights = unit.c(unit(1, "lines"), unit(1, "null")))

col2_with_title <- arrangeGrob(textGrob("Naïve vs Exhausted", gp = gpar(fontsize = 12, fontface = "bold")),
                               col2,
                               ncol = 1,
                               heights = unit.c(unit(1, "lines"), unit(1, "null")))

col3_with_title <- arrangeGrob(textGrob("Memory vs Exhausted", gp = gpar(fontsize = 12, fontface = "bold")),
                               col3,
                               ncol = 1,
                               heights = unit.c(unit(1, "lines"), unit(1, "null")))

# Combine label column and heatmap columns into full layout
grid.arrange(label_grobs, col1_with_title, col2_with_title, col3_with_title,
             ncol = 4,
             widths = unit.c(unit(1.5, "cm"), unit(1, "null"), unit(1, "null"), unit(1, "null")),
             top = "hPMT Profiles Across T-cell States")

```
The heatmaps reveal that each histone modification exhibits substantial and coordinated changes in signal intensity between T-cell states. Activating marks (H3K27ac, H3K4me3) generally show increased signal in memory T-cells compared to naïve, while repressive marks (H3K27me3, H3K9me3) are more prominent in exhausted T-cells. Hierarchical clustering demonstrates clear separation between cell types for all histone modifications, underscoring robust and reproducible epigenetic reprogramming during T-cell differentiation and exhaustion. These patterns suggest that both activation and repression of chromatin regions are key features distinguishing naïve, memory, and exhausted T-cell states.


## 6.5) Comparing all significant hPMTs of different cell type comparisons

To get an overview of how the different cell types differ in terms of hPMTs, all hPMT per cell type comparison were added to a new matrix. If a peak present in one hPMT was missing in another, values were set to zero for visualisatio and later removal.

### 6.5.1) Memory vs Naïve

```{r}
scaled_list <- list(scaled_signal_H3K4me3, scaled_signal_H3K27ac, scaled_signal_H3K27me3, scaled_signal_H3K9me3)

all_regions <- Reduce(union, lapply(scaled_list, rownames))

fill_missing_rows <- function(mat, all_rows) {
  missing <- setdiff(all_rows, rownames(mat))
  if(length(missing) > 0) {
    # Create zero matrix for missing rows
    zero_mat <- matrix(0, nrow = length(missing), ncol = ncol(mat),
                       dimnames = list(missing, colnames(mat)))
    # Bind existing matrix and zero rows
    mat <- rbind(mat, zero_mat)
  }
  # Reorder rows
  mat[all_rows, , drop = FALSE]
}
scaled_filled <- lapply(scaled_list, fill_missing_rows, all_rows = all_regions)

combined_scaled <- do.call(cbind, scaled_filled)

#rm(scaled_list)
```

```{r}
samples_H3K4me3  <- colnames(scaled_filled[[1]])
samples_H3K27ac  <- colnames(scaled_filled[[2]])
samples_H3K27me3 <- colnames(scaled_filled[[3]])
samples_H3K9me3  <- colnames(scaled_filled[[4]])

all_samples <- c(samples_H3K27ac, samples_H3K4me3, samples_H3K27me3, samples_H3K9me3)


annotation_col <- data.frame(
  CellType = rep(c("Naïve T-cell", "Memory T-cell"), times = c(4,4)),
  HistoneMark = rep(c( "H3K27ac", "H3K4me3", "H3K27me3", "H3K9me3"), each = 8)
)

annotation_colors <- list(
  CellType = c("Naïve T-cell" = "#ef77b4", "Memory T-cell" = "#ff7f0e"),
  HistoneMark = c(
  "H3K4me3"   = "#ffd92f",  
  "H3K27ac"   = "#D73027",  
  "H3K27me3"  = "#a6cee3",  
  "H3K9me3"   = "#6A5ACD"   
))


rownames(annotation_col) <- colnames(combined_scaled)

top_rows <- head(order(apply(combined_scaled, 1, var), decreasing = TRUE), 5000)
```


```{r}
all_hmap1 <- pheatmap::pheatmap(combined_scaled[top_rows, ], breaks = my_breaks, 
         annotation_col = annotation_col,
         annotation_colors = annotation_colors,
         show_colnames = FALSE,
         show_rownames = FALSE,
         main = "A) Naïve vs. Memory T-cell",
         cluster_rows = TRUE,
         cluster_cols = F,
         silent = TRUE,
         cellheight = 0.0275)

```


### 6.5.2)Exhausted vs Naïve 

```{r}
# List of all scaled matrices
scaled_list_EXN <- list(scaled_signal_H3K4me3_EXN, scaled_signal_H3K27ac_EXN, scaled_signal_H3K27me3_EXN, scaled_signal_H3K9me3_EXN)

all_regions_EXN <- Reduce(union, lapply(scaled_list_EXN, rownames))

fill_missing_rows <- function(mat, all_rows) {
  missing <- setdiff(all_rows, rownames(mat))
  if(length(missing) > 0) {
    # Create zero matrix for missing rows
    zero_mat <- matrix(0, nrow = length(missing), ncol = ncol(mat),
                       dimnames = list(missing, colnames(mat)))
    # Bind existing matrix and zero rows
    mat <- rbind(mat, zero_mat)
  }
  # Reorder rows
  mat[all_rows, , drop = FALSE]
}

scaled_filled_EXN <- lapply(scaled_list_EXN, fill_missing_rows, all_rows = all_regions_EXN)
combined_scaled_EXN <- do.call(cbind, scaled_filled_EXN)

```

```{r}
# Get colnames for each scaled matrix (samples)
samples_H3K4me3  <- colnames(scaled_filled_EXN[[1]])
samples_H3K27ac  <- colnames(scaled_filled_EXN[[2]])
samples_H3K27me3 <- colnames(scaled_filled_EXN[[3]])
samples_H3K9me3  <- colnames(scaled_filled_EXN[[4]])

all_samples_EXN <- c(samples_H3K27ac, samples_H3K4me3, samples_H3K27me3, samples_H3K9me3)

annotation_col_2 <- data.frame(
  CellType = rep(c("Naïve T-cell", "Exhausted T-cell"), times = c(4,4)),
  HistoneMark = rep(c( "H3K27ac", "H3K4me3", "H3K27me3", "H3K9me3"), each = 8)
)

annotation_colors_2 <- list(
  CellType = c("Naïve T-cell" = "#ef77b4", "Exhausted T-cell" = "#4daf4a"),
  HistoneMark = c(
  "H3K4me3"   = "#ffd92f",  
  "H3K27ac"   = "#D73027",  
  "H3K27me3"  = "#a6cee3",  
  "H3K9me3"   = "#6A5ACD"   
))
rownames(annotation_col_2) <- colnames(combined_scaled_EXN)


top_rows_EXN <- head(order(apply(combined_scaled_EXN, 1, var), decreasing = TRUE), 5000)

all_hmap2 <- pheatmap::pheatmap(combined_scaled_EXN[top_rows_EXN, ], breaks = my_breaks, 
         annotation_col = annotation_col_2,
         annotation_colors = annotation_colors_2,
         show_colnames = FALSE,
         show_rownames = FALSE,
         main = "B) Exhausted vs. Naïve T-cell",
         cluster_rows = TRUE,
         cluster_cols = F,
         silent = TRUE,
         cellheight = 0.0275)

#rm(top_rows, scaled_filled_EXN)
```

### 6.5.3) Memory vs Exhausted

```{r}
scaled_list_EXM <- list(scaled_signal_H3K4me3_EXM, scaled_signal_H3K27ac_EXM, scaled_signal_H3K27me3_EXM, scaled_signal_H3K9me3_EXM)

all_regions_EXM <- Reduce(union, lapply(scaled_list_EXM, rownames))

fill_missing_rows <- function(mat, all_rows) {
  missing <- setdiff(all_rows, rownames(mat))
  if(length(missing) > 0) {
    # Create zero matrix for missing rows
    zero_mat <- matrix(0, nrow = length(missing), ncol = ncol(mat),
                       dimnames = list(missing, colnames(mat)))
    # Bind existing matrix and zero rows
    mat <- rbind(mat, zero_mat)
  }
  # Reorder rows
  mat[all_rows, , drop = FALSE]
}

scaled_filled_EXM <- lapply(scaled_list_EXM, fill_missing_rows, all_rows = all_regions_EXM)
combined_scaled_EXM <- do.call(cbind, scaled_filled_EXM)
```


```{r}
samples_H3K4me3  <- colnames(scaled_filled_EXM[[1]])
samples_H3K27ac  <- colnames(scaled_filled_EXM[[2]])
samples_H3K27me3 <- colnames(scaled_filled_EXM[[3]])
samples_H3K9me3  <- colnames(scaled_filled_EXM[[4]])

all_samples_EXM <- c(samples_H3K27ac, samples_H3K4me3, samples_H3K27me3, samples_H3K9me3)



annotation_col_3 <- data.frame(
  CellType = rep(c("Memory T-cell", "Exhausted T-cell"), times = c(4,4)),
  HistoneMark = rep(c( "H3K27ac", "H3K4me3", "H3K27me3", "H3K9me3"), each = 8))

annotation_colors_3 <- list(
  CellType = c("Memory T-cell" = "#ff7f0e","Exhausted T-cell"= "#4daf4a"),
  HistoneMark = c(
  "H3K4me3"   = "#ffd92f",  
  "H3K27ac"   = "#D73027",  
  "H3K27me3"  = "#a6cee3",  
  "H3K9me3"   = "#6A5ACD"   
)
)

rownames(annotation_col_3) <- colnames(combined_scaled_EXM)

top_rows_EXM <- head(order(apply(combined_scaled_EXM, 1, var), decreasing = TRUE), 5000)

all_hmap3 <- pheatmap::pheatmap(combined_scaled_EXM[top_rows_EXM, ], breaks = my_breaks, 
         annotation_col = annotation_col_3,
         annotation_colors = annotation_colors_3,
         show_colnames = FALSE,
         show_rownames = FALSE,
         main = "C) Memory vs. Exhausted T-cell",
         cluster_rows = TRUE,
         cluster_cols = F,
         silent = TRUE,
         cellheight = 0.0275)

#rm(top_rows, scaled_filled_EXM)
```


```{r, fig.width=8, fig.height=9, fig.align = "center", fig.cap="Figure 5. Heatmaps of scaled signal intensities for differentially modified peaks across T-cell state transitions. Z-scored signal intensities are displayed for genomic regions with significant differences (|log₂ fold change| > 0.58496) between A) naïve vs. memory T-cells, B) exhausted vs. naïve T-cells, and C) memory vs. exhausted T-cells. Each panel shows signals across four histone modifications: H3K4me3, H3K27ac, H3K27me3, and H3K9me3. Rows represent individual genomic peaks, columns represent biological replicates. Color annotations indicate histone mark (top bar) and cell type (bottom bar). Hierarchical clustering reveals distinct chromatin modification patterns associated with T-cell differentiation and exhaustion states. "}
grid.arrange(
  all_hmap1$gtable, 
  all_hmap2$gtable, 
  all_hmap3$gtable, 
  ncol = 1,
  top = "Comparative Analysis of Histone Modification Landscapes Across T-cell Differentiation States"
)
```

This comparison illustrates a progressive epigenetic landscape transformation form naïve to memory and then exhausted T-cell states. Each transition shows distinct clustering patterns, which become clear when clustered according to the H3K27ac. 

#### Naïve vs. Memory
Enhanced activating marks (H3K4me3, H3K27ac) in memory cells support their "poised" chromatin state for rapid reactivation.
#### Naïve vs. Exhausted
Exhausted cells show increased repressive marks (H3K27me3, H3K9me3), consistent with their functionally suppressed phenotype
#### Memory vs. Exhausted
The most dramatic differences occur between these functionally opposite states, with memory cells maintaining active chromatin while exhausted cells accumulate repressive modifications


These results suggest that T-cell exhaustion represents an active epigenetic silencing process rather than simply a loss of memory cell characteristics. The distinct chromatin landscapes likely contribute to the functional differences between these T-cell states and may represent therapeutic targets for reversing T-cell exhaustion.

# 7.) Gene Annotation
For annotating genes we followed the transcription start site mapping done in the paper, of +/- 250kb, to evaluate distant enhancers and repressors, for active promoters we chose to use a more narrow window of 3000bp
```{r}
anno_H3K27ac <- annotatePeak(peaks$H3K27ac,
                             TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,
                             annoDb = "org.Mm.eg.db",
                             tssRegion = c(-250000, 250000),
                             verbose = FALSE)

anno_H3K27me3 <- annotatePeak(peaks$H3K27me3,
                              TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,
                              annoDb = "org.Mm.eg.db",
                              tssRegion = c(-250000, 250000),
                              verbose = FALSE)

anno_H3K4me3 <- annotatePeak(peaks$H3K4me3,
                              TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,
                              annoDb = "org.Mm.eg.db",
                              tssRegion = c(-250000, 250000),
                              verbose = FALSE)

anno_H3K9me3 <- annotatePeak(peaks$H3K9me3,
                              TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene,
                              annoDb = "org.Mm.eg.db",
                              tssRegion = c(-3000, 3000),
                              verbose = FALSE)
H3K27ac_anno_df <- as.data.frame(anno_H3K27ac) 
H3K27me3_anno_df <- as.data.frame(anno_H3K27me3) 
H3K4me3_anno_df <- as.data.frame(anno_H3K4me3)
H3K9me3_anno_df <- as.data.frame(anno_H3K9me3)

anno_1 <- H3K27ac_anno_df[, c("peak","SYMBOL")]
anno_2 <- H3K27me3_anno_df[, c("peak","SYMBOL")]
anno_3 <- H3K4me3_anno_df[, c("peak","SYMBOL")]
anno_4 <- H3K9me3_anno_df[, c("peak","SYMBOL")]

combined_symbols <- unique(rbind(anno_1, anno_2, anno_3, anno_4))
symbol_lookup <- setNames(combined_symbols$SYMBOL, combined_symbols$peak)
```

## 7.1) Annotation Preprocessing

For annotating genes we chose to exlude all rows that contain a 0 to hopefully have heatmaps that are a bit more easy to read.

```{r}
anno_MN <-combined_scaled[top_rows, ]
anno_EXN <-combined_scaled_EXN[top_rows_EXN, ]
anno_EXM <-combined_scaled_EXM[top_rows_EXM, ]

gene_symbols <- symbol_lookup[rownames(anno_MN)]
anno_MN_df <- as.data.frame(anno_MN)
anno_MN_clean <- anno_MN
rownames(anno_MN_clean) <- make.unique(as.character(gene_symbols))

gene_symbols <- symbol_lookup[rownames(anno_EXN)]
anno_EXN_df <- as.data.frame(anno_EXN)
anno_EXN_clean <- anno_EXN
rownames(anno_EXN_clean) <- make.unique(as.character(gene_symbols))

gene_symbols <- symbol_lookup[rownames(anno_EXM)]
anno_EXM_df <- as.data.frame(anno_EXM)
anno_EXM_clean <- anno_EXM
rownames(anno_EXM_clean) <- make.unique(as.character(gene_symbols))

anno_MN_filtered  <- anno_MN_clean[apply(anno_MN_clean != 0, 1, all), ,drop = FALSE]
dim(anno_MN_filtered)
anno_EXN_filtered  <- anno_EXN_clean[apply(anno_EXN_clean != 0, 1, all), ,drop = FALSE]
dim(anno_EXN_filtered)
anno_EXM_filtered  <- anno_EXM_clean[apply(anno_EXM_clean != 0, 1, all), ,drop = FALSE]
dim(anno_EXM_filtered)
```

## 7.2) Annotations of heatmaps
The first three matrices all contain 5000 peaks, since we previously filtered for the top 5000 peaks in every hPMT. Post filtering out any rows that contained a zero, i.e. not significant across all hPMTs, the datasets were substantially smaller.

For visualisation and readability of the heatmaps, we filtered the datasets again to only include the top 50 hits.

```{r}
top_rows_1<- head(order(apply(anno_MN_filtered, 1, var), decreasing = TRUE), 50)
annot_hmap1 <- pheatmap::pheatmap(anno_MN_filtered[top_rows_1, ], breaks = my_breaks, 
         annotation_col = annotation_col,
         annotation_colors = annotation_colors,
         show_colnames = FALSE,
         show_rownames = TRUE,
         main = "Naive vs. Memory T-cell",
         cluster_rows = TRUE,
         cluster_cols = F,
         silent = T,
         fontsize_row = 7,
         cellheight = 8,
         cellwidth = 10)

top_rows_2 <- head(order(apply(anno_EXN_filtered, 1, var), decreasing = TRUE), 50)
annot_hmap2 <- pheatmap::pheatmap(anno_EXN_filtered[top_rows_2, ], breaks = my_breaks, 
         annotation_col = annotation_col_2,
         annotation_colors = annotation_colors_2,
         show_colnames = FALSE,
         show_rownames = TRUE,
         main = "Naive vs. Exhausted T-cell",
         cluster_rows = TRUE,
         cluster_cols = F,
         silent = T,
         fontsize_row = 7,
         cellheight = 8,
         cellwidth = 10)

top_rows_3 <- head(order(apply(anno_EXM_filtered, 1, var), decreasing = TRUE), 50)
annot_hmap3 <- pheatmap::pheatmap(anno_EXM_filtered[top_rows_3, ], breaks = my_breaks, 
         annotation_col = annotation_col_3,
         annotation_colors = annotation_colors_3,
         show_colnames = FALSE,
         show_rownames = TRUE,
         main = "Memory vs. Exhausted T-cell",
         cluster_rows = TRUE,
         cluster_cols = F,
         silent = T,
         fontsize_row = 7,
         cellheight = 8,
         cellwidth = 10)

```


```{r, fig.width=8, fig.height=20, fig.align='center', fig.cap = "Figure 6. Heatmaps of Differential Histone Modification Signals with Zero-Value Peaks Removed. Each heatmap displays z-scored signal intensities for regions with significant differences (|log₂ fold change| > 0.58496) across four histone modifications (H3K4me3, H3K27ac, H3K27me3, H3K9me3) in pairwise comparisons: naïve vs. memory, naïve vs. exhausted, and memory vs. exhausted T-cells. Only peaks with non-zero signal across all samples in each comparison are included. Rows represent genomic peaks; columns represent biological replicates, annotated by histone mark and cell type. "}
grid.arrange(
  annot_hmap1$gtable, 
  annot_hmap2$gtable, 
  annot_hmap3$gtable, 
  ncol = 1,
  top = "Heatmaps of Top 50 Differential Histone Modification Signals with Gene Annotations"
)
```
These heatmaps display the top 50 differential histone modification signals for each comparison between memory and exhausted T-cells, grouped by gene and annotated by histone mark and cell type. Without delving into the specific gene annotations, several key points can be made:
*Clear Epigenetic Differences*: The heatmaps reveal distinct patterns of hPMTs (H3K4me3, H3K27ac, H3K27me3, H3K9me3) between naive, memory, and exhausted T-cells, visible as clusters of genes with coordinated changes in modification signal intensity.
Cell Type Segregation: The annotation bars indicate that samples cluster according to cell type, suggesting that memory and exhausted T-cells have systematically different chromatin landscapes at these loci.
*Histone Mark-Specific Patterns*: The color-coded annotation for histone marks shows that different marks contribute variably to the observed differences, with some clusters enriched for activating marks (H3K4me3, H3K27ac) and others for repressive marks (H3K27me3, H3K9me3).
*Basis for Downstream Analysis*: This heatmap panel serves as a visual summary of the most epigenetically dynamic genes between cell states, justifying the selection of these genes for subsequent functional analyses.


# 8.) Numerical Quantification of Heatmaps
## 8.1) Numerical evaluation of activating, repressing and bivalent regions across cell type comparisons
Next to the visual evaluation done using the heatmaps above, we were also interested in quantifying the high signal activating, repressing, and bivalent hPMT's across our cell type comparisons.

### 8.1.1) Naïve and Memory 

```{r}

H3K4me3_MN  <- combined_scaled[, 5:8]
H3K27ac_MN  <- combined_scaled[, 13:16]
H3K27me3_MN <- combined_scaled[, 21:24]
H3K9me3_MN  <- combined_scaled[, 29:32]

# Activating High
open_high_H3K4me3_MN  <- apply(H3K4me3_MN, 1, function(x) all(x >= 0))
open_high_H3K27ac_MN  <- apply(H3K27ac_MN, 1, function(x) all(x >= 0))

select_rows_MN <- which(open_high_H3K4me3_MN & open_high_H3K27ac_MN)

activating_MN <- combined_scaled[select_rows_MN, c(5:8, 13:16, 21:24, 29:32)]
activating_MN <- activating_MN[!apply(activating_MN == 0, 1, any), ]


write.csv(activating_MN, file = "files/open_peaks_signal.csv", row.names = T)
```

```{r}
# Repressed High
closed_high_H3K27me3_MN <- apply(H3K27me3_MN, 1, function(x) all(x >= 0))
closed_high_H3K9me3_MN  <- apply(H3K9me3_MN, 1, function(x) all(x >= 0))

select_rows_rep_MN <- which(closed_high_H3K27me3_MN & closed_high_H3K9me3_MN)

repressing_MN <- combined_scaled[select_rows_rep_MN, c(5:8, 13:16, 21:24, 29:32)]
repressing_MN <- repressing_MN[!apply(repressing_MN == 0, 1, any), ]


write.csv(repressing_MN, file= "files/closed_peaks_signal.csv", row.names = T)
```

```{r}
# Bivalent 

select_rows_b_MN <- which(open_high_H3K4me3_MN & closed_high_H3K27me3_MN)

bivalent_MN <- combined_scaled[select_rows_b_MN, c(5:8, 13:16, 21:24, 29:32)]
bivalent_MN <- bivalent_MN[!apply(bivalent_MN == 0, 1, any), ]


write.csv(bivalent_MN, file = "files/bivalent.csv", row.names = T)
```
Using this approach we decrease the size of the dataset for comparing naïve and memory T-cells to 308 peaks, which is substantially less than the initial 45174 peaks we looked at.

### 8.1.2) Naïve and Exhausted
```{r}
# Activating High
H3K4me3_EXN  <- combined_scaled_EXN[, 5:8]
H3K27ac_EXN  <- combined_scaled_EXN[, 13:16]
H3K27me3_EXN <- combined_scaled_EXN[, 21:24]
H3K9me3_EXN  <- combined_scaled_EXN[, 29:32]

open_high_H3K4me3_EXN  <- apply(H3K4me3_EXN, 1, function(x) all(x >= 0))
open_high_H3K27ac_EXN  <- apply(H3K27ac_EXN, 1, function(x) all(x >= 0))

select_rows_EXN <- which(open_high_H3K4me3_EXN & open_high_H3K27ac_EXN) 
activating_EXN <- combined_scaled_EXN[select_rows_EXN, c(5:8, 13:16, 21:24, 29:32)]
activating_EXN <- activating_EXN[!apply(activating_EXN == 0, 1, any), ]


write.csv(activating_EXN, file = "files/open_peaks_signal_EXN.csv", row.names = T)
```

```{r}
# Repressing High
closed_high_H3K27me3_EXN <- apply(H3K27me3_EXN, 1, function(x) all(x >= 0))
closed_high_H3K9me3_EXN  <- apply(H3K9me3_EXN, 1, function(x) all(x >= 0))


select_rows_rep_EXN <- which(closed_high_H3K27me3_EXN & closed_high_H3K9me3_EXN) 

repressing_EXN <- combined_scaled_EXN[select_rows_rep_EXN, c(5:8, 13:16, 21:24, 29:32)]
repressing_EXN <- repressing_EXN[!apply(repressing_EXN == 0, 1, any), ]


write.csv(repressing_EXN, file= "files/closed_peaks_signal_EXN.csv", row.names = T)
```


```{r}
# Bivalent
select_rows_b_EXN <- which(open_high_H3K4me3_EXN & closed_high_H3K27me3_EXN)
bivalent_EXN <- combined_scaled_EXN[select_rows_b_EXN, c(5:8, 13:16, 21:24, 29:32)]
bivalent_EXN <- bivalent_EXN[!apply(bivalent_EXN == 0, 1, any), ]



write.csv(bivalent_EXN, file = "files/bivalent_EXN.csv", row.names = T)
```


### 8.1.3) Exhausted and Memory
```{r}
# Activating High
H3K4me3_EXM  <- combined_scaled_EXM[, 5:8]
H3K27ac_EXM  <- combined_scaled_EXM[, 13:16]
H3K27me3_EXM <- combined_scaled_EXM[, 21:24]
H3K9me3_EXM  <- combined_scaled_EXM[, 29:32]

open_high_H3K4me3_EXM  <- apply(H3K4me3_EXM, 1, function(x) all(x >= 0))
open_high_H3K27ac_EXM  <- apply(H3K27ac_EXM, 1, function(x) all(x >= 0))

select_rows_EXM <- which(open_high_H3K4me3_EXM & open_high_H3K27ac_EXM) 

activating_EXM <- combined_scaled_EXM[select_rows_EXM, c(5:8, 13:16, 21:24, 29:32)]
activating_EXM <- activating_EXM[!apply(activating_EXM == 0, 1, any), ]


write.csv(activating_EXM, file = "files/open_peaks_signal_EXM.csv", row.names = T)
```

```{r}
# Repressing High

closed_high_H3K27me3_EXM <- apply(H3K27me3_EXM, 1, function(x) all(x >= 0))
closed_high_H3K9me3_EXM  <- apply(H3K9me3_EXM, 1, function(x) all(x >= 0))

select_rows_rep_EXM <- which(closed_high_H3K27me3_EXM & closed_high_H3K9me3_EXM) 

repressing_EXM <- combined_scaled[select_rows_rep_EXM, c(5:8, 13:16, 21:24, 29:32)]
repressing_EXM <- repressing_EXM[!apply(repressing_EXM == 0, 1, any), ]

write.csv(repressing_EXM, file= "files/closed_peaks_signal_EXM.csv", row.names = T)
```


```{r}
# Bivalent
select_rows_b_EXM <- which(open_high_H3K4me3_EXM & closed_high_H3K27me3_EXM)
bivalent_EXM <- combined_scaled[select_rows_b_EXM, c(5:8, 13:16, 21:24, 29:32)]
bivalent_EXM <- bivalent_EXM[!apply(bivalent_EXM == 0, 1, any), ]

write.csv(bivalent_EXM, file = "files/bivalent_EXM.csv", row.names = T)
```


```{r}
comparison <- matrix(nrow = 3, ncol = 3)
colnames(comparison) <- c("Naïve vs. Memory", "Naïve vs. Exhausted",  "Memory vs. Exhausted" )
rownames(comparison) <- c("activating", "repressing", "bivalent")

comparison[1,1] <- nrow(activating_MN)
comparison[2,1] <-nrow(repressing_MN)
comparison[3,1] <-nrow(bivalent_MN)

comparison[1,2] <-nrow(activating_EXN)
comparison[2,2] <-nrow(repressing_EXN)
comparison[3,2] <-nrow(bivalent_EXN)


comparison[1,3] <-nrow(activating_EXM)
comparison[2,3] <-nrow(repressing_EXM)
comparison[3,3] <-nrow(bivalent_EXM)
```


```{r, tab.cap="Table 2. Comparing the counts of significant activating, repressing, and bivalent peaks across cell type comparisons"}
comparison
```
Table 2 summarises the counts of significant activating, repressing, and bivalent chromatin peaks identified in each pairwise comparison of T-cell states. In the transition from naïve to memory T-cells, 131 activating, 141 repressing, and 148 bivalent peaks were detected. The comparison between naïve and exhausted T-cells revealed higher numbers of significant peaks, with 179 activating, 237 repressing, and 196 bivalent regions. Notably, the memory versus exhausted comparison showed a marked increase in repressing peaks (560), while the number of activating (3) and bivalent (105) peaks was substantially lower. These results indicate that chromatin repression is most pronounced during the transition from memory to exhausted T-cells, whereas both activation and bivalency are more prominent in earlier stages of T-cell differentiation.


```{r, warning=T, include = F, eval=F}
#rm(list = ls())
```
